main.cpp

```diff
@@ -12,147 +12,367 @@
 //    See the License for the specific language governing permissions and
 //    limitations under the License.
 
-#include <iostream>
-
-#include <libflash/fileio.hpp>
-#include <libflash/optimized_writer.hpp>
-#include <common/common.hpp>
+#define _GNU_SOURCE /* needed for splice() */
 
+#include <errno.h>
+#include <fcntl.h>
 #include <getopt.h>
+#include <mtd/ubi-user.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <sys/sendfile.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <unistd.h>
+
+#include "config.h"
+
+#define UBIMajorDevNo 10
+#define BLOCK_SIZE (1024 * 1024L) /* 1 MiB */
+#define MIN(X, Y) ((X < Y) ? X : Y)
+
+static struct option long_options[] = {
+	{"help", no_argument, 0, 'h'},
+	{"write-everything", no_argument, 0, 'w'},
+	{"input-size", required_argument, 0, 's'},
+	{"fsync-interval", required_argument, 0, 'f'},
+	{"input", required_argument, 0, 'i'},
+	{"output", required_argument, 0, 'o'},
+	{0, 0, 0, 0}};
 
 void PrintHelp() {
-	std::cout
-		<< "Usage: mender-flash [-h|--help] [-s|--input-size <INPUT_SIZE>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>"
-		<< std::endl;
+	fputs(
+		"Usage:\n"
+		"  mender-flash [-h|--help] [-w|--write-everything] [-s|--input-size <INPUT_SIZE>] [-f|--fsync-interval <FSYNC_INTERVAL>] -i|--input <INPUT_PATH> -o|--output <OUTPUT_PATH>\n",
+		stderr);
 }
 
-int main(int argc, char *argv[]) {
-	size_t volumeSize = 0;
-	std::string inputPath;
-	std::string outputPath;
-	const size_t blockSize = 1024 * 1024; // 1MiB block size
-
-	while (true) {
-		static struct option long_options[] = {
-			{"help", no_argument, 0, 'h'},
-			{"input-size", required_argument, 0, 's'},
-			{"input", required_argument, 0, 'i'},
-			{"output", required_argument, 0, 'o'},
-			{0, 0, 0, 0}};
-
-		int option_index = 0;
-		int c = getopt_long(argc, argv, "hs:i:o:", long_options, &option_index);
-		if (c == -1) {
-			break;
+typedef ssize_t (*io_fn_t)(int, void *, size_t);
+
+struct Stats {
+	size_t blocks_written;
+	size_t blocks_omitted;
+	uint64_t bytes_written;
+	uint64_t bytes_omitted;
+	uint64_t total_bytes;
+};
+
+ssize_t buf_io(io_fn_t io_fn, int fd, unsigned char *buf, size_t len) {
+	size_t rem = len;
+	ssize_t n_done;
+	do {
+		n_done = io_fn(fd, buf + (len - rem), rem);
+		if (n_done > 0) {
+			rem -= n_done;
+		} else if ((n_done == -1) && (errno == EINTR)) {
+			continue;
 		}
+	} while ((n_done > 0) && (rem > 0) && (len > 0));
+
+	if (n_done < 0) {
+		return n_done;
+	} else {
+		return (len - rem);
+	}
+}
+
+bool shovel_data(
+	int in_fd,
+	int out_fd,
+	size_t len,
+	bool write_optimized,
+	size_t fsync_interval,
+	struct Stats *stats,
+	int *error) {
```

Comment:
I noticed there isn't much use of `const`. What is the general opinion / recommendation?

In other projects we like to use const basically whenever possible. I found this in the Google C++ style guide:

https://google.github.io/styleguide/cppguide.html#Use_of_const

> We strongly recommend using const in APIs (i.e., on function parameters, methods, and non-local variables) wherever it is meaningful and accurate.

> Using const on local variables is neither encouraged nor discouraged.

So I'd consider adding const here and in other places, but you might have a different opinion on what should / shouldn't be, so I won't comment line by line.

Reply:
There are no `const *` candidates. Using `const` on arguments in the function **definition** has always looked weird to me. And a too frequent of `const` on local variables, IMHO, results in `const` kind of meaning it's benefit of actually explicitly pointing out that something should really be constant in some block. But YMMV.